packages/flame/lib/src/experimental/geometry/shapes/polygon.dart

```diff
@@ -0,0 +1,219 @@
+import 'dart:ui';
+
+import 'package:vector_math/vector_math_64.dart';
+
+import '../../../game/transform2d.dart';
+import 'shape.dart';
+
+/// An arbitrary polygon with 3 or more vertices.
+///
+/// The vertices of the polygon are in the counter-clockwise order, so that the
+/// polygon's interior is in your left-hand side as you traverse the polygon's
+/// vertices.
+///
+/// A polygon can be either convex or not, the [containsPoint] method will work
+/// in both cases, however, the method used with convex polygon is faster.
+class Polygon extends Shape {
+  /// Constructs the polygon from the given list of vertices.
+  ///
+  /// If the list is not in the counter-clockwise order, then it will be
+  /// reversed in-place.
+  ///
+  /// If the [convex] flag is provided, then it serves as a hint about whether
+  /// the polygon is convex or not. With this flag the user promises that the
+  /// vertices are already in the correct CCW order.
+  Polygon(this._vertices, {bool? convex})
+      : assert(_vertices.length >= 3, 'At least 3 vertices are required') {
+    _initializeEdges();
+    if (convex == null) {
+      _checkOrientation();
+    } else {
+      _convex = convex;
+    }
+  }
+
+  /// The vertices (corners) of the polygon.
+  ///
+  /// The user should treat this list as read-only and not attempt to modify
+  /// either the list itself or individual points.
+  List<Vector2> get vertices => _vertices;
+  final List<Vector2> _vertices;
+
+  /// The edges (sides) of the polygon.
+  ///
+  /// Each i-th edge is equal to the vector difference between the i-th vertex
+  /// and the preceding vertex. The number of edges is always equal to the
+  /// number of vertices.
+  List<Vector2> get edges => _edges;
+  late List<Vector2> _edges;
+  void _initializeEdges() {
+    var previousVertex = _vertices.last;
+    _edges = _vertices.map((Vector2 vertex) {
+      final edge = vertex - previousVertex;
+      previousVertex = vertex;
+      return edge;
+    }).toList(growable: false);
+  }
+
+  /// Number of vertices/edges in the polygon.
+  int get n => _vertices.length;
```

Comment:
This should have a more descriptive name

Reply:
I was basing this on the fact that a [polygon](https://en.wikipedia.org/wiki/Polygon) is also called an **n-gon**, so the property _n_ of a polygon is actually the standard terminology.

Reply:
Sounds like something that would be more known to matematicians, but maybe not programmers? I haven't seen any geometrical game api that exposes it as n at least, but maybe you have some examples?

Reply:
I think something like `numberOfSides` would be clearer (albeit much longer), that is what they use in the text on the wikipedia page too.

Reply:
That's even longer than `edges.length` though. Maybe just remove it?

Reply:
True, I think `edges.length` is nice and clear

Reply:
ok, i removed `.n`.